[JENKINS-63587] Allow disabling Ant default excludes in the zip step

ykhandelwal913 opened the pull request:
Bydefault zipcreation excludes the default files like .gitignore and others mentioned  in https://ant.apache.org/manual/dirtasks.html#defaultexcludes.

This PR is to solve https://issues.jenkins.io/browse/JENKINS-63587

Example:
`zip zipFile: 'template.zip', archive: true, defaultExcludes:false, glob: 'template-*/**/*'`
default value of defaultExcludes is true.

rsandell reviewed (changes requested):
Missing documentation

rsandell commented:
See https://github.com/jenkinsci/pipeline-utility-steps-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep and https://github.com/jenkinsci/pipeline-utility-steps-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/tar/TarStep

Also if this only applies to zip files then the parameter should be added to that step, not the abstract step that also is used by `TarStep`.

ykhandelwal913 commented:
> See https://github.com/jenkinsci/pipeline-utility-steps-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep and https://github.com/jenkinsci/pipeline-utility-steps-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/tar/TarStep
> 
> Also if this only applies to zip files then the parameter should be added to that step, not the abstract step that also is used by `TarStep`.

It is only applies to zip files.

ykhandelwal913 commented:
> Missing documentation

Done. Added in code itself. is that fine?

rsandell commented:
You also need to add a `help-defaultExcludes.html` to the resource directory https://github.com/jenkinsci/pipeline-utility-steps-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep with user documentation explaining what it is about, not every user can read java code :)

ykhandelwal913 commented:
@rsandell not sure why build is failing. Can you help?

rsandell commented on src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep/help-defaultExclude.html:
```diff
@@ -0,0 +1,28 @@
+  -->
+
+<p>
+    <a href="https://ant.apache.org/manual/dirtasks.html#defaultexcludes" target="_blank">Ant style pattern</a>
```
This doesn't seem to be correct. It's a checkbox, how is the user supposed to provide an ant style pattern?

rsandell commented on src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/AbstractFileCompressStep.java:
```diff
@@ -114,4 +115,25 @@ public void setOverwrite(boolean overwrite) {
+     *
+     * @return the defaultExcludes boolean
+     */
+    public boolean isDefaultExcludes() {
```
Because this feature only applies to Zip files, and the configuration options is only in that class, this code should be moved into the `ZipStep` class.

ykhandelwal913 replied:
Moved it to ZipStep class. Can you review please.

basil commented:
Looks like the build failure is caused by a Javadoc error:

```
16:49:16  [ERROR] /home/jenkins/agent/workspace/line-utility-steps-plugin_PR-180/src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/AbstractFileCompressStep.java:132: error: invalid use of @return
16:49:16  [ERROR]      * @return the defaultExcludes boolean
16:49:16  [ERROR]        ^
```

basil commented:
PS You can test Javadoc generation locally by running `mvn clean verify -DskipTests -Pjenkins-release`.

ykhandelwal913 commented:
> Looks like the build failure is caused by a Javadoc error:
> 
> ```
> 16:49:16  [ERROR] /home/jenkins/agent/workspace/line-utility-steps-plugin_PR-180/src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/AbstractFileCompressStep.java:132: error: invalid use of @return
> 16:49:16  [ERROR]      * @return the defaultExcludes boolean
> 16:49:16  [ERROR]        ^
> ```

Thanks for the pointer. Fixed it.

basil commented on src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStepExecution.java:
```diff
@@ -75,10 +75,13 @@ static class ZipItFileCallable extends AbstractFileCallable<Integer> {
         final boolean overwrite;
 
-        public ZipItFileCallable(String glob, String exclude, boolean overwrite) {
+        private boolean defaultExcludes = true;
```
Making this field `final` would make it both consistent with the other fields in this class and easier to reason about (mutable state is generally harder to reason about than immutable state).

ykhandelwal913 replied:
@basil If i make it final, how can i assign the default value as true as by default we want not to includes patterns which are excludes by ant java pattern.

ykhandelwal913 replied:
Done. can you review?

basil commented on src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep/help-defaultExclude.html:
```diff
@@ -0,0 +1,28 @@
+  ~ SOFTWARE.
+  -->
+
+<p>
+    The zip file should include defaultExcludes provided by ant java pattern.
+    By Default defaultExcludes param value will be true.
+</p>
```
This documentation is grammatically confusing (the first sentence probably meant to say "_Whether_ the zip file should […]") and also has incorrect capitalization. Suggest clarifying as follows:

```
<p>By default, the following exclusions are applied:</p>

<pre>**/*~
**/#*#
**/%*%
**/.#*
**/._*
**/.bzr
**/.bzr/**
**/.bzrignore
**/CVS
**/CVS/**
**/.cvsignore
**/.DS_Store
**/.git
**/.git/**
**/.gitattributes
**/.gitignore
**/.gitmodules
**/.hg
**/.hg/**
**/.hgignore
**/.hgsub
**/.hgsubstate
**/.hgtags
**/SCCS
**/SCCS/**
**/.svn
**/.svn/**
**/vssver.scc
</pre>

<p>Set this option to <code>false</code> if you do not want these default exclusions to be applied.</p>
```

basil commented on src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep.java:
```diff
@@ -44,6 +45,29 @@
+    private boolean defaultExcludes = true;
+
+    /**
+     * <a href="https://ant.apache.org/manual/dirtasks.html#defaultexcludes" target="_blank">Ant style pattern</a>
+     * of files to disable default excludes from the archive.
```
Incorrect: this is a boolean, not a pattern.

basil commented on src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep.java:
```diff
@@ -44,6 +45,29 @@
+    }
+
+    /**
+     * <a href="https://ant.apache.org/manual/dirtasks.html#defaultexcludes" target="_blank">Ant style pattern</a>
+     * of files to disable default excludes from the archive.
```
Incorrect: this is a boolean, not a pattern.

basil commented:
> It is only applies to zip files.

From my reading of the code, it appears the tar step uses the same Ant task and therefore the same default exclusions. If true, then it seems inconsistent to add this parameter to the unzip step but not the tar step.

ykhandelwal913 commented:
> From my reading of the code, it appears the tar step uses the same Ant task and therefore the same default exclusions. If true, then it seems inconsistent to add this parameter to the zip step but not the tar step.

As this ticket was for zip step i added code only for that but i agree with your point and to have the same consistency in zip and tar, added the same in tarStep as well.

ykhandelwal913 commented:
@basil can this be reviewed now. This feature is much needed for us to progress in our project. Hence the urgency.

basil commented on src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/ZipStep.java:
```diff
@@ -34,6 +34,7 @@
 import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
 import org.jenkinsci.plugins.workflow.steps.StepExecution;
 import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
```
I do not see a reason to add an import that is not used. This seems like leftover code from an earlier revision. Suggest removing the addition of this import.

ykhandelwal913 replied:
Done.

basil reviewed (commented):
This is looking good overall, but I think now that the scope of this PR extends to both the zip and the tar step, HTML documentation for the tar step needs to be added to match that of the zip step, along with a corresponding test for the tar step. Otherwise the change is inconsistent, applying some logic to one step but not the other.

basil reviewed (approved):
Looks good to me. Thanks for your contribution and for addressing feedback so quickly! In order to set expectations properly, I am not a maintainer of this plugin (though I am a Jenkins core maintainer), so I will be unable to merge your change. But @rsandell is a maintainer of this plugin, and I believe that the work we have done together has brought this PR closer to a state that can be approved by him (if not to the finish line!). Thanks again!

ykhandelwal913 commented:
> Looks good to me. Thanks for your contribution and for addressing feedback so quickly! In order to set expectations properly, I am not a maintainer of this plugin (though I am a Jenkins core maintainer), so I will be unable to merge your change. But @rsandell is a maintainer of this plugin, and I believe that the work we have done together has brought this PR closer to a state that can be approved by him (if not to the finish line!). Thanks again!

Thanks for the feedback and support provided.

rsandell commented:
Looks good, from what I can see you only need to fix the javadoc errors in the new code to make the build pass.

ykhandelwal913 commented:
@rsandell can this be merged?

ykhandelwal913 commented:
It didnot create a new release. Do i need to do anything from my side?
